README
$Id: README,v 1.26 2010/03/19 19:28:18 ianmacd Exp $


Introduction
------------

Ruby/AWS is a Ruby language library that aims to make it relatively easy for
the programmer to retrieve information from the popular Amazon Web site via
Amazon's Associates Web Services (AWS). In addition to the original amazon.com
site, the local sites amazon.co.uk, amazon.de, amazon.fr, amazon.ca and
amazon.co.jp are also supported.

Development of Ruby/AWS has been quite swift since the appearance of the first
alpha version, 0.0.1, in late March of 2008. Although Ruby/AWS shares almost no
code with its now obsolete predecessor, Ruby/Amazon, many lessons were learnt
whilst developing that library, and the experience gained has been rolled into
Ruby/AWS.

As of version 0.3.0, I believe that Ruby/AWS has attained its goal of being
superior to the final version of Ruby/Amazon, 0.9.2, which was released in
August 2006.


History and compatibility with Ruby/Amazon
------------------------------------------

In the beginning, there was Ruby/Amazon. This library was built around version
3.x of the Amazon Web Service API and first saw the light of day in January
2004. The version of the Amazon API in use at the time was known as AWS 3.x.

Amazon later renamed AWS to ECS, or E-Commerce Service, for the launch of
version 4 of their API, a complete overhaul that provided no backward
compatibility with previous versions. The previous version of the API was
thenceforth sometimes referred to as ECS 3.

Demonstrating the wisdom and consistency for which large companies are
renowned, Amazon changed their mind once again in late 2007, reverting to the
familiar name of AWS. This time, however, it was said to stand for Associates
Web Service, rather than Amazon Web Service.

Since Amazon first made AWS available, the number of Amazon Web APIs has
grown and AWS is now just one of many. It is therefore no longer appropriate
to call this library by a name so general as Ruby/Amazon, because it
provides an interface to just one of the Amazon Web APIs. Therefore, the
monicker for this library is Ruby/AWS.

Unfortunately for Ruby/AWS, Amazon changed the name once again in May 2009,
referring to it now as the Product Advertising API. Changing Ruby/AWS's name
would create more confusion than it would mitigate, however, so I'm not about
to do so. Similarly, I will continue to refer to the Amazon API in question as
AWS.

Ruby/AWS is built around version 4 of the Amazon AWS API, which is
fundamentally different to version 3, both in terms of how requests are made
and the data returned. The underlying structure of the XML response has
radically changed from previous versions.

It has therefore not been practical for Ruby/AWS to retain any level of API
compatibility with Ruby/Amazon. Unfortunately, this means that any code
written for Ruby/Amazon will need to be rewritten to work with Ruby/AWS. The
good news is that, in most cases, this isn't as much work as it might sound.

Another bit of good news is that the /etc/amazonrc and ~/.amazonrc files used
by Ruby/Amazon _are_ compatible with Ruby/AWS. The only change required for
Ruby/AWS is the addition of the 'key_id' and 'secret_key_id' parameters, which
should contain your AWS Access Key ID and its secret counterpart. That fact
notwithstanding, as of version 0.5.0, Ruby/AWS also supports a more flexible,
locale-specific configuration syntax.

Amazon finally decomissioned v3 of the AWS API on 2008-03-31. As a result, the
original Ruby/Amazon library no longer functions and is therefore obsolete.


AWS Access Key ID
-----------------

You can obtain an AWS Access Key ID here:

  https://aws-portal.amazon.com/gp/aws/developer/registration/index.html

You may see mention of Subscription IDs at the above location. Subscription
IDs are not supported by Ruby/AWS and, in any case, are no longer supported by
Amazon since the introduction of authenticated requests. Please obtain and use
an AWS Access Key ID instead.


API version
-----------

Ruby/AWS currently requests the 2009-11-01 revision of the AWS API when
performing its operations:

  http://docs.amazonwebservices.com/AWSECommerceService/2009-11-01/DG/

However, a different version can be requested via the 'api' parameter in the
user configuration file.


Status and functionality
------------------------

Ruby/AWS is currently beta software. Amongst other things, this means:

- You will encounter bugs, but hopefully not too many and none too serious.
  Tell me about them and I will endeavour to fix them.

- The documentation isn't what it could be, but it's hopefully enough to get
  you up and running.

- Not all features are currently implemented. Others may not yet be _fully_
  implemented. Some, I probably haven't even thought of yet. Again, if
  something's missing, tell me, and if it makes sense, I'll add it.

  In spite of this shortcomings, the AWS v4 API is more or less fully
  supported, with only small gaps in the functionality of some operations.

  Currently implemented operations are:
  
    BrowseNodeLookup
    CustomerContentLookup
    CustomerContentSearch
    Help
    ItemLookup
    ItemSearch
    ListLookup
    ListSearch
    SellerListingLookup
    SellerListingSearch
    SellerLookup
    SimilarityLookup
    TagLookup
    TransactionLookup

  Remote shopping-carts are also implemented as of version 0.3.0. The
  following remote shopping-cart operations are supported:

    CartCreate
    CartAdd
    CartModify
    CartClear

  Version 0.4.0 added the remaining shopping-cart operation, which I had first
  thought superfluous:

    CartGet

  Multiple operations are supported, but have not been heavily used, so there
  may be residual bugs. The implementation of multiple operations was
  rewritten from scratch for version 0.8.0, which appears to have resulted in
  code that is much more robust (not to mention easier to read and maintain).

  As of version 0.5.0, batch operations are fully supported, using the
  Operation#batch method. Again, the implementation was scrapped and rewritten
  for 0.8.0, yielding a big improvement.

  There appear to also be (undocumented) Amazon-imposed restrictions on the
  use of multiple operations and batch requests, so some experimentation may
  be required on your part to determine what works and what doesn't.
  
  In short, though, my investigations have demonstrated that no more than two
  operations may be batched together, and no more than two operations may be
  combined in a multiple operation. Nevertheless, by combining two batched
  operations in a multiple operation, one can send four base operations to
  Amazon in a single request. No more than two of these may be from the same
  class, however.

  The 2008-08-19 version of the AWS API added the following operations:

    VehiclePartLookup
    VehiclePartSearch
    VehicleSearch

  These are supported by Ruby/AWS from version 0.5.0 onwards.
  
- Classes, methods, constants and instance variables may change name in the
  future. New ones may appear from nowhere and existing ones may change shape,
  grow, shrink or disappear without trace. Such fundamental changes will break
  existing code, so I will endeavour to keep them to a minimum.

In short, code written to work with this release of Ruby/AWS may stop working
when you upgrade to the next. In fact, it may even stop working _during_ this
release cycle, because it's possible there are fatal conditions that I didn't
encounter in my limited testing of the code. It's also possible that future
(possibly unannounced) changes made by Amazon will affect Ruby/AWS in
ways I can't anticipate.

That said, the Ruby/AWS API is pretty stable at this point in time. I won't
break any of the method interfaces without seriously considering the merits of
doing so.


Installation
------------

Please see the INSTALL file for details of how to install Ruby/AWS. You can
choose between an installation script and a RubyGems installation.


Usage
-----

First of all, create either /etc/amazonrc or ~/.amazonrc as a plain text file.
Its contents should look something like this:

  # Any line that starts with a hash character is a comment.
  key_id = '0Y44V8G41KCQPGF6XYZ2'
  secret_key_id = 'k+kuddeoQJzUnImC0Hyy21J4xLWQc1hbvfQ+7F1G
  associate = 'fuzbarorg-21'
  cache = false
  locale = 'uk'
  encoding = 'iso-8859-15'

The ability to include your secret key was a feature added in version 0.6.0 of
Ruby/AWS. If you choose to do so, your requests to AWS will be signed for
authentication by Amazon's servers. If you don't, your requests will fail,
because Amazon made this practice obligatory on 15th August 2009.

When Amazon checks for a valid signature, it does so by comparing its
computation of the signature with the one supplied by the user. In doing this,
Amazon uses the UTF-8 representation of parameter values that you supply, even
if you use a different encoding.

In order to have Ruby/AWS properly reencode your strings as UTF-8, you need
to tell it which encoding you are using. The 'encoding' parameter can be used
for this, but you can omit it if your strings are already UTF-8, because this
is the default.

As of version 0.5.0 of Ruby/AWS, the following locale-specific configuration
syntax is also supported:

  [global]
  key_id = '0Y44V8G41KCQPGF6XYZ2'
  secret_key_id = 'k+kuddeoQJzUnImC0Hyy21J4xLWQc1hbvfQ+7F1G
  cache = false
  locale = 'uk'
  encoding = 'iso-8859-15'
  # Request a specific version of the API.
  # api = '2008-03-03'

  [uk]
  associate = 'fuzbarorg-21'

  [us]
  associate = 'fuzbarorg-20'

This enables the use of a different associate tag for each locale, although,
oddly enough, no-one has ever requested this feature.

Because you're embedding your personal keys in the file, you should protect it
(on UNIX and equivalent systems) by making it mode 0600:

$ chmod 600 ~/.amazonrc

If you define 'cache' to be 'true', you may want to also define 'cache_dir' to
point to somewhere other the default cache directory, /tmp/amazon.

If you want to place .amazonrc somewhere other than $HOME, you may set
the $AMAZONRCDIR environment variable, as this location is checked prior to
$HOME.

If you're using Windows, $HOME is usually undefined, so a number of additional
locations are checked for .amazonrc.

The exact search order is as follows:

$AMAZONRCDIR
$HOME
$HOMEDRIVE + $HOMEPATH
$USERPROFILE

Note that only the first defined location is used, so if, for example, both
$AMAZONRCDIR and $HOME are defined, but only the path specified by $HOME
contains a file called .amazonrc, it will not be found.

If you want the user configuration file to be called something other than
.amazonrc, you may define the $AMAZONRCFILE environment variable.

Once you have your configuration file, you can get started writing your code.
 
Here's a basic example, showing how to perform an ItemSearch, probably the
most common type of AWS operation. Please see the ./examples subdirectory for
more examples of contrived, but working code.

--

require 'amazon/aws/search'

# Avoid having to fully qualify our methods.
#
include Amazon::AWS
include Amazon::AWS::Search

is = ItemSearch.new( 'Books', { 'Title' => 'Ruby' } )

# I want to receive just a small amount of data for the items found.
#
is.response_group = ResponseGroup.new( :Small )

req = Request.new

# Make sure I'm talking to amazon.co.uk.
#
req.locale = 'uk'

# Actually talk to AWS.
#
resp = req.search( is )

# Drill down to the meat: the array of items returned.
#
items = resp.item_search_response[0].items[0].item

# The following alternative shorthand would also have worked:
#
# items = resp.item_search_response.items.item

# Available properties for first item:
#
puts items[0].properties

items.each do |item|
  attribs = item.item_attributes[0]
  puts attribs.label
  if attribs.list_price
    puts attribs.title, attribs.list_price[0].formatted_price, ''
  end
end


Troubleshooting
---------------

HTTP 400 is the main bane of people's life since Amazon started authenticating
requests to AWS. If you're trying to get Ruby/AWS to work, but are plagued by
HTTP 400 responses, here are some possible causes:

- Your ~/.amazonrc file doesn't contain a secret_key_id parameter. Add one.

- Your computer's system clock is running more than 15 minutes slow. Amazon
  consider a request timestamped more than 15 minutes in the past invalid.
  Synchronise your system clock with a reliable time source.

  Incidentally, Amazon don't object to requests timestamped in the future, but
  that's something that may change at any time and therefore shouldn't be
  relied upon.


XML to Ruby mapping
-------------------

Here, I will discuss the mapping of the XML returned from AWS to native Ruby
objects and data. Note that the XML shown below was that returned at the time
of writing and may look different to what you would see today if you were to
execute the same request.

When this code:

  resp = req.search( is )

was called in the previous section, the following URL was composed and sent to
AWS as an HTTP GET operation:

http://ecs.amazonaws.co.uk/onca/xml?AWSAccessKeyId=01234567890123456789&AssociateTag=calibanorg-21&Operation=ItemSearch&ResponseGroup=Small&SearchIndex=Books&Service=AWSECommerceService&Title=Ruby&Version=2008-03-03

The following (abbreviated) AWS XML response was received:

<ItemSearchResponse>
  <OperationRequest>
    <HTTPHeaders>
      <Header Name="UserAgent" Value="Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.8.1.13) Gecko/20080325 Fedora/2.0.0.13-1.fc7 Firefox/2.0.0.13"/>
    </HTTPHeaders>
    <RequestId>1TBGEZ48MF8KZ8TGXH65</RequestId>
    <Arguments>
      <Argument Name="SearchIndex" Value="Books"/>
      <Argument Name="Service" Value="AWSECommerceService"/>
      <Argument Name="ResponseGroup" Value="Small"/>
      <Argument Name="Operation" Value="ItemSearch"/>
      <Argument Name="Version" Value="2008-03-03"/>
      <Argument Name="AssociateTag" Value="calibanorg-21"/>
      <Argument Name="Title" Value="Ruby"/>
      <Argument Name="AWSAccessKeyId" Value="01234567890123456789"/>
    </Arguments>
    <RequestProcessingTime>0.0671439170837402</RequestProcessingTime>
  </OperationRequest>
  <Items>
    <Request>
      <IsValid>True</IsValid>
      <ItemSearchRequest>
        <ResponseGroup>Small</ResponseGroup>
        <SearchIndex>Books</SearchIndex>
        <Title>Ruby</Title>
      </ItemSearchRequest>
    </Request>
    <TotalResults>1804</TotalResults>
    <TotalPages>181</TotalPages>
    <Item>
      <ASIN>0439943663</ASIN>
      <DetailPageURL>
http://www.amazon.co.uk/gp/redirect.html%3FASIN=0439943663%26tag=calibanorg-21%26lcode=xm2%26cID=2025%26ccmID=165953%26location=/o/ASIN/0439943663%253FSubscriptionId=0Y44V8FAFNM119C6PTR2
      </DetailPageURL>
      <ItemAttributes>
        <Author>Philip Pullman</Author>
        <Manufacturer>Scholastic</Manufacturer>
        <ProductGroup>Book</ProductGroup>
        <Title>The Ruby in the Smoke (Sally Lockhart Quartet)</Title>
      </ItemAttributes>
    </Item>
    <Item>
      <ASIN>0596516177</ASIN>
      ...

In Ruby/AWS, each unique XML element name forms a class of the same name. All
such classes are subclasses of AWSObject. For example, OperationRequest is a
class, as is ItemAttributes.

As the XML tree is traversed, each element is converted to an instance of the
class of the same name. Every such object has instance variables, one per
unique child element name. The name of the instance variable is translated to
comply with Ruby convention by adding an underscore ('_') character at word
boundaries and converting the name to lower case.

For example, given the following XML:

<ItemAttributes>
   <Author>Philip Pullman</Author>
   <Manufacturer>Scholastic</Manufacturer>
   <ProductGroup>Book</ProductGroup>
   <Title>The Ruby in the Smoke (Sally Lockhart Quartet)</Title>
</ItemAttributes>

the following statements would all be true:

- ItemAttributes, Author, Manufacturer, ProductGroup and Title would all be
  dynamically defined subclasses of AWSObject.

- An instance of the ItemAttributes class would be created, with instance

NEWS
$Id: NEWS,v 1.21 2010/03/20 11:56:30 ianmacd Exp $


0.8.1 - 2010-03-20
------------------

1. Operation#query_parameters was faulty, because it had accidentally been
   defined outside the Operation class.

   This meant that it wasn't in the correct namespace, causing libraries that
   included Ruby/AWS to place that method in a namespace unreachable by
   Ruby/AWS itself. Scripts that used Ruby/AWS without defining other modules
   and classes were unaffected.

2. In determining the home directory location of the .amazonrc file, we were
   mistakenly concatenating ENV['HOMEDRIVE'] + ENV['HOMEPATH'] without first
   checking that both were defined.

3. The version of the AWS API used is now 2009-11-01, the latest at the time
   of writing.


0.8.0 - 2010-02-21
------------------

1. There is a new exception class, ObsolescenceError, which is raised when
   an obsolete (i.e. no longer available) feature is used.

2. The batching and multiple operation code was basically unmaintainable. I
   could no longer comprehend it, so I removed and rewrote it from scratch.

3. The implementation of MultipleOperation had a bug, whereby, if the second
   operation was a batched operation, its parameters would overwrite those of
   the first operation, leading to too few results returned.
   
   This meant that the maximum number of operations that could be encapsulated
   in a MultipleOperation was 3, not the proper 4, and, even then, only when
   the batched operation was supplied as the first parameter, not the second.

4. There's a new ResponseGroup#response_group= method, the use of which is
   transparent to the user. This makes it possible to assign response groups
   to any kind of operation, even those encapsulated in batched and multiple
   operations.

5. MultipleOperation.new can now accept an arbitrary number of operations in
   its parameter list. This list may also be supplied in the form of an array.

   Amazon still impose a maximum of 2 operations encapsulated in a multiple
   operation, but there's no reason for Ruby/AWS to mirror this. If Amazon
   later raise the limit, Ruby/AWS will require no modification to take
   advantage of the new situation.

6. It's now possible to pass two batched operations to MultipleOperation.new,
   allowing 4 base operations to be sent to AWS in a single request.
   
   Note that Amazon still impose a limit of no more than 2 operations of the
   same class in a single request, which means that to achieve the maximum of
   4 base operations in a MultipleOperation, one must encapsulate two batched
   operations of different classes.

7. A MultipleOperation that encapsulates only operations of the same class
   will be commuted to a batch operation, resulting in the appropriate
   response from Amazon.
   
   For example, a MultipleOperation encapsulating two ItemSearch operations
   will return an item_search_response, not a multiple_operation_response. A
   MultipleOperation encapsulating an ItemSearch and an ItemLookup, on the
   other hand, will still return a multiple_operation_response.
   
   Previously, all multiple operations returned a multiple_operation_response.

8. It's no longer possible to pass a list of response groups as the second
   parameter to Search::Request#search. This facility was deprecated in
   Ruby/AWS 0.7.0 and has been removed altogether in 0.8.0 (shunting nr_pages
   into place as the second and final parameter).
   
   You must now directly assign to the operation's @response_group attribute.

   In the case of a batched operation or a MultipleOperation, the response
   groups will automatically propagate to the encapsulated operations.

   An attempt to pass response groups as the second parameter to
   Search::Request#search will now yield an ObsolescenceError exception.

9. Attempting to retrieve ALL_PAGES of a multiple page list, such as a
   WishList, didn't work. Only the first page was retrieved. This is because
   TotalPages is nested one level deeper in the XML response yielded by this
   type of operation. The same is true of MultipleOperation responses.

10. Geolocation now associates Belgian clients with amazon.fr, rather than
    amazon.co.uk. A Belgian client may not necessarily be a Francophone, but
    this is probably still an improvement.

11. The version of the AWS API used is now 2009-10-01, the latest at the time
    of writing. ItemSearch::SEARCH_INDICES has been updated with the latest
    additions, Lighting and Outlet.

12. Ruby/AWS now depends on Ruby 1.8.6, not 1.8.7. This should make life
    easier for some people, as many Linux distributions still come with 1.8.6.


0.7.0 - 2009-06-16
------------------

1. This release introduces a shorthand module method for each of the AWS
   operations. These can be used to create less verbose code at the expense of
   flexibility.

   For example, we might normally write the following code:

     is = ItemSearch.new( 'Books', { 'Title' => 'Ruby' } )
     rg = ResponseGroup.new( :Large )
     req = Request.new
     response = req.search( is, rg )
     
   but we could instead use ItemSearch's associated module method as follows:

     response = Amazon::AWS.item_search( 'Books', { 'Title' => 'Ruby' } )

   There are some important restrictions when compared to the standard way of
   doing things:

   a. Astute readers will note that there's no way to specify to the module
      methods which response group(s) to use. Instead, a reasonable default
      set for each type of operation will be used, as per the new
      ResponseGroup::DEFAULT hash.

      The exception to this is Amazon::AWS.multiple_operation, which has no
      response groups of its own, instead applying those of the operations it
      combines.

      Because no access is provided to the Request object used by the module
      method, it's also not possible to batch operations with Operation#batch
      when using this form of the search.


lib/amazon/aws/search.rb
# $Id: search.rb,v 1.49 2010/03/19 19:28:19 ianmacd Exp $
#

module Amazon

  module AWS

    require 'amazon/aws'
    require 'net/http'
    require 'rexml/document'
    require 'openssl'

    # Load this library with:
    #
    #  require 'amazon/aws/search'
    #
    module Search

      class Request

	include REXML

	# Exception class for bad access key ID.
	#
	class AccessKeyIdError < Amazon::AWS::Error::AWSError; end

	# Exception class for bad locales.
	#
	class LocaleError < Amazon::AWS::Error::AWSError; end

	# Do we have support for the SHA-256 Secure Hash Algorithm?
	#
	# Note that Module#constants returns Strings in Ruby 1.8 and Symbols
	# in 1.9.
	#
	DIGEST_SUPPORT = OpenSSL::Digest.constants.include?( 'SHA256' ) ||
			 OpenSSL::Digest.constants.include?( :SHA256 )

	# Requests are authenticated using the SHA-256 Secure Hash Algorithm.
	#
	DIGEST = OpenSSL::Digest::Digest.new( 'sha256' ) if DIGEST_SUPPORT

	attr_reader :conn, :config, :locale, :query, :user_agent
	attr_writer :cache
	attr_accessor :encoding

	# This method is used to generate an AWS search request object.
	#
	# _key_id_ is your AWS {access key
	# ID}[https://aws-portal.amazon.com/gp/aws/developer/registration/index.html].
	# Note that your secret key, used for signing requests, can be
	# specified only in your <tt>~/.amazonrc</tt> configuration file.
	#
	# _associate_ is your
	# Associates[http://docs.amazonwebservices.com/AWSECommerceService/2009-11-01/GSG/BecominganAssociate.html]
	# tag (if any), _locale_ is the locale in which you which to work
	# (*us* for amazon.com[http://www.amazon.com/], *uk* for
	# amazon.co.uk[http://www.amazon.co.uk], etc.), _cache_ is whether or
	# not you wish to utilise a response cache, and _user_agent_ is the
	# client name to pass when performing calls to AWS. By default,
	# _user_agent_ will be set to a string identifying the Ruby/AWS
	# library and its version number.
	#
	# _locale_ and _cache_ can also be set later, if you wish to change
	# the current behaviour.
	#
	# Example:
	#
	#  req = Request.new( '0Y44V8FAFNM119CX4TR2', 'calibanorg-20' )
	#
	def initialize(key_id=nil, associate=nil, locale=nil, cache=nil,
		       user_agent=USER_AGENT)

	  @config ||= Amazon::Config.new

	  def_locale = locale
	  locale = 'us' unless locale
	  locale.downcase!

	  key_id ||= @config['key_id']
	  cache = @config['cache'] if cache.nil?

	  # Take locale from config file if no locale was passed to method.
	  #
	  if @config.key?( 'locale' ) && ! def_locale
	    locale = @config['locale']
	  end
	  validate_locale( locale )

	  if key_id.nil?
	    raise AccessKeyIdError, 'key_id may not be nil'
	  end

	  @key_id     = key_id
	  @tag	      = associate || @config['associate'] || DEF_ASSOC[locale]
	  @user_agent = user_agent
	  @cache      = unless cache == 'false' || cache == false
			  Amazon::AWS::Cache.new( @config['cache_dir'] )
			else
			  nil
			end

	  # Set the following two variables from the config file. Will be
	  # *nil* if not present in config file.
	  #
	  @api	      = @config['api']
	  @encoding   = @config['encoding']

	  self.locale = locale
	end


	# Assign a new locale. If the locale we're coming from is using the
	# default Associate ID for that locale, then we use the new locale's
	# default ID, too.
	#
	def locale=(l)  # :nodoc:
	  old_locale = @locale ||= nil
	  @locale = validate_locale( l )

	  # Use the new locale's default ID if the ID currently in use is the
	  # current locale's default ID.
	  #
	  if @tag == Amazon::AWS::DEF_ASSOC[old_locale]
	    @tag = Amazon::AWS::DEF_ASSOC[@locale]
	  end

	  if @config.key?( @locale ) && @config[@locale].key?( 'associate' )
	    @tag = @config[@locale]['associate']
	  end

	  # We must now set up a new HTTP connection to the correct server for
	  # this locale, unless the same server is used for both.
	  #
	  unless Amazon::AWS::ENDPOINT[@locale] ==
		 Amazon::AWS::ENDPOINT[old_locale]
	    #connect( @locale )
	    @conn = nil
	  end
	end


	# If @cache has simply been assigned *true* at some point in time,
	# assign a proper cache object to it when it is referenced. Otherwise,
	# just return its value.
	#
	def cache  # :nodoc:
	  if @cache == true
	    @cache = Amazon::AWS::Cache.new( @config['cache_dir'] )
	  else
	    @cache
	  end
	end


	# Verify the validity of a locale string. _l_ is the locale string.
	#
	def validate_locale(l)
	  unless Amazon::AWS::ENDPOINT.has_key? l
	    raise LocaleError, "invalid locale: #{l}"
	  end
	  l
	end
	private :validate_locale


	# Return an HTTP connection for the current _locale_.
	#
	def connect(locale)
	  if ENV.key? 'http_proxy'
	    uri = URI.parse( ENV['http_proxy'] )
	    proxy_user = proxy_pass = nil
	    proxy_user, proxy_pass = uri.userinfo.split( /:/ ) if uri.userinfo
	    @conn = Net::HTTP::Proxy( uri.host, uri.port, proxy_user,
				      proxy_pass ).start(
					Amazon::AWS::ENDPOINT[locale].host )
	  else
	    @conn = Net::HTTP::start( Amazon::AWS::ENDPOINT[locale].host )
	  end
	end
	private :connect


	# Reconnect to the server if our connection has been lost (due to a
	# time-out, etc.).
	#
	def reconnect  # :nodoc:
	  connect( self.locale )
	  self
	end


	# This method checks for errors in an XML response returned by AWS.
	# _xml_ is the XML node below which to search.
	#
	def error_check(xml)
	  if ! xml.nil? && xml = xml.elements['Errors/Error']
	    raise Amazon::AWS::Error.exception( xml )
	  end
	end
	private :error_check


	# Add a timestamp to a request object's query string.
	#
	def timestamp # :nodoc:
	  @query << '&Timestamp=%s' %
	    [ Amazon.url_encode(
		Time.now.utc.strftime( '%Y-%m-%dT%H:%M:%SZ' ) ) ]
	end
	private :timestamp


	# Add a signature to a request object's query string. This implicitly
	# also adds a timestamp.
	#
	def sign # :nodoc:
	  return false unless DIGEST_SUPPORT

	  timestamp
	  params = @query[1..-1].split( '&' ).sort.join( '&' )
  
	  sign_str = "GET\n%s\n%s\n%s" % [ ENDPOINT[@locale].host,
					   ENDPOINT[@locale].path,
					   params ]

	  Amazon.dprintf( 'Calculating SHA256 HMAC of "%s"...', sign_str )

	  hmac = OpenSSL::HMAC.digest( DIGEST,
				       @config['secret_key_id'],
				       sign_str )
	  Amazon.dprintf( 'SHA256 HMAC is "%s"', hmac.inspect )

	  base64_hmac = [ hmac ].pack( 'm' ).chomp
	  Amazon.dprintf( 'Base64-encoded HMAC is "%s".', base64_hmac )

	  signature = Amazon.url_encode( base64_hmac )

	  params << '&Signature=%s' % [ signature ]
	  @query = '?' + params

	  true
	end


	# Perform a search of the AWS database, returning an AWSObject.
	#
	# _operation_ is an object of a subclass of _Operation_, such as
	# _ItemSearch_, _ItemLookup_, etc. It may also be a _MultipleOperation_
	# object.
	#
	# In versions of Ruby/AWS up to prior to 0.8.0, the second parameter to
	# this method was _response_group_. This way of passing response
	# groups has been deprecated since 0.7.0 and completely removed in
	# 0.8.0. To pair a set of response groups with an operation, assign
	# directly to the operation's @response_group attribute.
	#
	# _nr_pages_ is the number of results pages to return. It defaults to
	# <b>1</b>. If a higher number is given, pages 1 to _nr_pages_ will be
	# returned. If the special value <b>:ALL_PAGES</b> is given, all
	# results pages will be returned.
	#
	# Note that _ItemLookup_ operations can use several different
	# pagination parameters. An _ItemLookup_ will typically return just
	# one results page containing a single product, but <b>:ALL_PAGES</b>
	# can still be used to apply the _OfferPage_ parameter to paginate
	# through multiple pages of offers.
	#
	# Similarly, a single product may have multiple pages of reviews
	# available. In such a case, it is up to the user to manually supply
	# the _ReviewPage_ parameter and an appropriate value.
	#
	# In the same vein, variations can be returned by using the
	# _VariationPage_ parameter.
	#
	# The pagination parameters supported by each type of operation,
	# together with the maximum page number that can be retrieved for each
	# type of data, are # documented in the AWS Developer's Guide:
	#
	# http://docs.amazonwebservices.com/AWSECommerceService/2009-11-01/DG/index.html?MaximumNumberofPages.html
	#
	# The pagination parameter used by <b>:ALL_PAGES</b> can be looked up
	# in the Amazon::AWS::PAGINATION hash.
	#
	# If _operation_ is of class _MultipleOperation_, the operations
	# encapsulated within will return only the first page of results,
	# regardless of whether a higher number of pages is requested.
	#
	# If a block is passed to this method, each successive page of results
	# will be yielded to the block.
	#
	def search(operation, nr_pages=1)
	  parameters = Amazon::AWS::SERVICE.
			 merge( { 'AWSAccessKeyId' => @key_id,
				  'AssociateTag'   => @tag } ).
			 merge( operation.query_parameters )

	  if nr_pages.is_a? Amazon::AWS::ResponseGroup
	    raise ObsolescenceError, 'Request#search method no longer accepts response_group parameter.'
	  end

	  # Pre-0.8.0 user code may have passed *nil* as the second parameter,
	  # in order to use the @response_group of the operation.
	  #
	  nr_pages ||= 1

	  # Check to see whether a particular version of the API has been
	  # requested. If so, overwrite Version with the new value.
	  #
	  parameters.merge!( { 'Version' => @api } ) if @api

	  @query = Amazon::AWS.assemble_query( parameters, @encoding )
	  page = Amazon::AWS.get_page( self )

	  # Ruby 1.9 needs to know that the page is UTF-8, not ASCII-8BIT.
	  #
	  page.force_encoding( 'utf-8' ) if RUBY_VERSION >= '1.9.0'

	  doc = Document.new( page )

	  # Some errors occur at the very top level of the XML. For example,
	  # when no Operation parameter is given. This should not be possible
	  # with user code, but occurred during debugging of this library.
	  #
	  error_check( doc )

	  # Another possible error results in a document containing nothing
	  # but <Result>Internal Error</Result>. This occurs when a specific
	  # version of the AWS API is requested, in combination with an
	  # operation that did not yet exist in that version of the API.
	  #
	  # For example:
	  #
	  # http://ecs.amazonaws.com/onca/xml?AWSAccessKeyId=foo&Operation=VehicleSearch&Year=2008&ResponseGroup=VehicleMakes&Service=AWSECommerceService&Version=2008-03-03
	  #
	  if xml = doc.elements['Result']
	    raise Amazon::AWS::Error::AWSError, xml.text
	  end

	  # Fundamental errors happen at the OperationRequest level. For
	  # example, if an invalid AWSAccessKeyId is used.
	  #
	  error_check( doc.elements['*/OperationRequest'] )

	  # Check for parameter and value errors deeper down, inside Request.
	  #
	  if operation.kind == 'MultipleOperation'

	    # Everything is a level deeper, because of the
	    # <MultiOperationResponse> container.
	    #
	    # Check for errors in the first operation.
	    #
	    error_check( doc.elements['*/*/*/Request'] )

	    # Check for errors in the second operation.
	    #
	    error_check( doc.elements['*/*[3]/*/Request'] )

	    # If second operation is batched, check for errors in its 2nd set
	    # of results.
	    #
	    if batched = doc.elements['*/*[3]/*[2]/Request']
	      error_check( batched )
	    end
	  else
	    error_check( doc.elements['*/*/Request'] )

	    # If operation is batched, check for errors in its 2nd set of
	    # results.
	    #
	    if batched = doc.elements['*/*[3]/Request']
	      error_check( batched )
	    end
	  end

	  if doc.elements['*/*[2]/TotalPages']
	    total_pages = doc.elements['*/*[2]/TotalPages'].text.to_i

	  # FIXME: ListLookup and MultipleOperation (and possibly others) have
	  # TotalPages nested one level deeper. I should take some time to
	  # ensure that all operations that can return multiple results pages
	  # are covered by either the 'if' above or the 'elsif' here.
	  #
	  elsif doc.elements['*/*[2]/*[2]/TotalPages']
	    total_pages = doc.elements['*/*[2]/*[2]/TotalPages'].text.to_i
	  else
	    total_pages = 1
	  end

	  # Create a root AWS object and walk the XML response tree.
	  #
	  aws = AWS::AWSObject.new( operation )
	  aws.walk( doc )
	  result = aws

	  # If only one page has been requested or only one page is available,
	  # we can stop here. First yield to the block, if given.
	  #
	  if nr_pages == 1 || ( tp = total_pages ) == 1
	     yield result if block_given?
	     return result
	  end

	  # Limit the number of pages to the maximum number available.
	  #
	  nr_pages = tp.to_i if nr_pages == :ALL_PAGES || nr_pages > tp.to_i

	  if PAGINATION.key? operation.kind
	    page_parameter = PAGINATION[operation.kind]['parameter']
	    max_pages = PAGINATION[operation.kind]['max_page']
	  else
	    page_parameter = 'ItemPage'
	    max_pages = 400
	  end

	  # Iterate over pages 2 and higher, but go no higher than MAX_PAGES.
	  #
	  2.upto( nr_pages < max_pages ? nr_pages : max_pages ) do |page_nr|
	    @query = Amazon::AWS.assemble_query(
		      parameters.merge( { page_parameter => page_nr } ),
		      @encoding)
	    page = Amazon::AWS.get_page( self )

	    # Ruby 1.9 needs to know that the page is UTF-8, not ASCII-8BIT.
	    #
	    page.force_encoding( 'utf-8' ) if RUBY_VERSION >= '1.9.0'

	    doc = Document.new( page )

	    # Check for errors.
	    #
	    error_check( doc.elements['*/OperationRequest'] )
	    error_check( doc.elements['*/*/Request'] )

	    # Create a new AWS object and walk the XML response tree.
	    #
	    aws = AWS::AWSObject.new( operation )
	    aws.walk( doc )

	    # When dealing with multiple pages, we return not just an
	    # AWSObject, but an array of them.
	    #
	    result = [ result ] unless result.is_a? Array

	    # Append the new object to the array.
	    #
	    result << aws
	  end

	  # Yield each object to the block, if given.
	  #
	  result.each { |r| yield r } if block_given?

	  result
	end

      end

    end

  end

end
